Source/wave.cpp

```diff
@@ -2,311 +2,221 @@
+	if (chunk == NULL)
+		return TRUE;
+	return ReadWaveSection(pMemFile, MAKEFOURCC('d','a','t','a'), chunk);
 }
 
-int __fastcall ReadMemFile(MEMFILE *pMemFile, void *lpBuf, size_t a3)
-{
-	size_t v3; // ebx
-	void *v4; // ebp
-	MEMFILE *v5; // esi
-	size_t v6; // edi
-
-	v3 = a3;
-	v4 = lpBuf;
-	v5 = pMemFile;
-	if ( !a3 )
-		return 1;
-	while ( 1 )
-	{
-		if ( !v5->bytes_to_read )
-			FillMemFile(v5);
-		v6 = v5->bytes_to_read;
-		if ( v3 < v6 )
-			v6 = v3;
-		if ( !v6 )
-			break;
-		memcpy(v4, &v5->buf[v5->dist], v6);
-		v5->offset += v6;
-		v5->dist += v6;
-		v5->bytes_to_read -= v6;
-		v3 -= v6;
-		if ( !v3 )
-			return 1;
+BOOL __fastcall ReadMemFile(MEMFILE *pMemFile, void *lpBuf, size_t length)
+{
+	while (length) {
+		size_t to_copy;
+		if (!pMemFile->bytes_to_read)
+			FillMemFile(pMemFile);
+		to_copy = pMemFile->bytes_to_read;
+		if (length < to_copy)
+			to_copy = length;
+		if (!to_copy)
+			return FALSE;
+		memcpy(lpBuf, &pMemFile->buf[pMemFile->dist], to_copy);
+		pMemFile->offset += to_copy;
+		pMemFile->dist += to_copy;
+		pMemFile->bytes_to_read -= to_copy;
+		// BUGFIX: lpBuf is not incremented, next read will overwrite data
+		length -= to_copy;
 	}
-	return 0;
+	return TRUE;
 }
 
 void __fastcall FillMemFile(MEMFILE *pMemFile)
 {
-	MEMFILE *v1; // esi
-	unsigned int v2; // edi
-
-	v1 = pMemFile;
-	WSetFilePointer((HANDLE)pMemFile->file, pMemFile->offset, 0, 0);
-	v2 = v1->end - v1->offset;
-	if ( v1->buf_len < v2 )
-		v2 = v1->buf_len;
-	if ( v2 )
-		WReadFile((HANDLE)v1->file, v1->buf, v2);
-	v1->dist = 0;
-	v1->bytes_to_read = v2;
+	DWORD to_read;
+	WSetFilePointer(pMemFile->file, pMemFile->offset, NULL, SEEK_SET);
+	to_read = pMemFile->end - pMemFile->offset;
+	if (pMemFile->buf_len < to_read)
+		to_read = pMemFile->buf_len;
+	if (to_read)
+		WReadFile(pMemFile->file, pMemFile->buf, to_read);
+	pMemFile->dist = 0;
+	pMemFile->bytes_to_read = to_read;
 }
 
-int __fastcall SeekMemFile(MEMFILE *pMemFile, unsigned int lDist, int dwMethod)
+int __fastcall SeekMemFile(MEMFILE *pMemFile, LONG lDist, DWORD dwMethod)
 {
-	unsigned int v3; // eax
-
-	v3 = pMemFile->bytes_to_read;
-	if ( lDist >= v3 )
-	{
-		pMemFile->bytes_to_read = 0;
-	}
-	else
+	if (lDist < pMemFile->bytes_to_read)
 	{
+		pMemFile->bytes_to_read -= lDist;
 		pMemFile->dist += lDist;
-		pMemFile->bytes_to_read = v3 - lDist;
 	}
+	else
+		pMemFile->bytes_to_read = 0;
 	pMemFile->offset += lDist;
 	return pMemFile->offset;
 }
 
-int __fastcall ReadWaveSection(MEMFILE *pMemFile, int a2, int *a3)
+BOOL __fastcall ReadWaveSection(MEMFILE *pMemFile, DWORD id, riff_section *chunk)
 {
-	int v3; // esi
-	MEMFILE *v4; // edi
-	int v6; // eax
-	int a2a[2]; // [esp+8h] [ebp-8h]
-
-	v3 = a2;
-	v4 = pMemFile;
-	while ( 1 )
-	{
-		if ( !ReadMemFile(v4, a2a, 8u) )
-			return 0;
-		if ( a2a[0] == v3 )
+	DWORD hdr[2];
+
+	do {
+		if (!ReadMemFile(pMemFile, hdr, sizeof(hdr)))
+			return FALSE;
+		if (hdr[0] == id)
 			break;
-		if ( SeekMemFile(v4, a2a[1], FILE_CURRENT) == -1 )

```

Comment:
`FILE_CURRENT` came from the assert strings in 1.00. Pretty sure that was used instead of `SEEK_CUR`

Reply:
Figures, why use libc when there's windows-specific junk...